### DISCUSSION TITLE
Is there any official tcp/udp hello world coding sample?

### DISCUSSION THREAD

**Discussion by krishna116** [libuv/libuv#3336] [Q&A]
Hi  everyone,

The code I found is either too old to run or have incorrect memory operation(alloc/free) and memory-leak dangerous.
I want  safe/up-to-date samples to specify:
  1, a minimal code how to using tcp/udp communication?
  2, where/when should I allocate memory for handle/request/buf?
  3, where/when should I free memory for handle/request/buf?

thank you.

**Comment by bnoordhuis** [comment]
There are plenty of examples in [the guide](https://docs.libuv.org/en/v1.x/guide.html), aren't there?

**Reply by krishna116** [reply]
for example, this tcp code is incomplete, it cannot run:
 https://docs.libuv.org/en/v1.x/guide/networking.html#tcp

**Comment by bnoordhuis** [comment]
Look here: https://github.com/libuv/libuv/tree/v1.x/docs/code

**Reply by krishna116** [reply]
it has a tcp echo server, but it doesn't have a  tcp echo client,
and I have already found a working client now, thank you all the same.
